File: Python3/Valid Parentheses.py
# 36ms, 100.0%
class Solution:
    def isValid(self, s):
        tmp, right, valid = s[0], ")]}", ["()", "[]", "{}"]
        for i in range(1, len(s)):
            if not tmp:
                if s[i] in right:
                    return False
                else:
                    tmp = s[i]
            elif tmp[-1] + s[i] in valid:
                tmp = tmp[:-1]
            else:
                tmp += s[i]
        return not tmp
    

File: Java/Valid Parentheses.java
// 3ms, 14.18%; 39.5MB, 5.06%
class Solution {
    public boolean isValid(String s) {
        if ((s.length() & 1) == 1) {
            return false;
        }
        Stack<Character> st = new Stack<Character>();
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            } else {
                char c2 = (c == ')' ? '(' : c == '}' ? '{' : '[');
                if (st.isEmpty() || st.lastElement() != c2) {
                    return false;
                }
                st.pop();
            }
        }
        return st.isEmpty();
    }
}
